genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,554 @@
+// required. (mkwinsyscall has defaults that may be acceptable.) It is recommended to read the
+// DllImport pseudo-attribute (§II.21.2.1) to determine this value.
+func (c *Context) WriteMethod(w io.StringWriter, method *winmd.MethodDef, moduleName, goName string) error {
+	w.WriteString("//sys\t")
+	writeEscapedUpper(w, goName)
+	w.WriteString("(")
+
+	sig, err := c.Metadata.MethodDefSignature(method.Signature)
+	if err != nil {
+		return err
+	}
+
+	for paramRowIndex := method.ParamList.Start; paramRowIndex < method.ParamList.End; paramRowIndex++ {
+		param, err := c.Metadata.Tables.Param.Record(paramRowIndex)
+		if err != nil {
+			return fmt.Errorf("failed to read param row %v defined by method %v: %w", paramRowIndex, method.Name, err)
+		}
+		if param.Sequence == 0 {
+			// Sequence 0 is ".param". A row with Sequence 0 is often included before other rows,
+			// but Sequence 0 doesn't contain any data in cases seen so far.
+			if param.Flags == 0 && param.Name.String() == "" {
+				continue
+			}
+			// TODO: Add support for assemblies that do have data in Sequence 0.
+			return fmt.Errorf("unsupported method: expected param row with sequence 0 to be empty, but: %#v", param)
+		}
+
+		// Param data is in two places: a param table row and the signature. Index into the
+		// signature param slice by converting the 1-based sequence value. Note that we assume
+		// ascending Sequence values for proper formatting: this is stated to be true in the
+		// informational section of §II.22.33, point 4.
+		// TODO: Handle gaps in Sequence values? §II.22.33 information point 5.
+		// Gaps have not been seen in Windows.Win32.winmd. The meaning of a gap is not yet known.
+		i := param.Sequence - 1
+		if i > 0 {
+			w.WriteString(", ")
+		}
+		c.writeParam(w, param)
+		w.WriteString(" ")
+
+		if int(i) >= len(sig.Param) {
+			return fmt.Errorf("param record Sequence value %v is out of range of parsed signature params, length %v", i, len(sig.Param))
+		}
+		if err := c.writeType(w, &sig.Param[i].Type); err != nil {
+			return fmt.Errorf("failed to interpret type of param %v of method %v: %w", i, method.Name, err)
+		}
+	}
+	w.WriteString(")")
+
+	// Write return value, if one exists.
+	if sig.RetType.Kind != winmd.SigRetTypeKind_Void {
+		w.WriteString(" (")
+		if err := c.writeType(w, &sig.RetType.Type); err != nil {
+			return err
+		}
+		w.WriteString(")")
+	}
+
+	// Write syscall name and module if non-defaults are needed.
+	if goName != method.Name.String() || moduleName != "" {
+		w.WriteString(" = ")
+		if moduleName != "" {
+			w.WriteString(moduleName)
+			w.WriteString(".")
+		}
+		writeEscapedUpper(w, method.Name.String())
+	}
+	return nil
+}
+
+func (c *Context) writeParam(w io.StringWriter, p *winmd.Param) {
+	// TODO: Use p.Flags to determine how to translate each param to a Go type?
+	writeEscapedParam(w, p.Name.String())
+}
+
+func (c *Context) writeType(b io.StringWriter, p *winmd.SigType) error {
```

Comment:
nitpick: I don't like defining internal helper functions like this one. `writeType` is a good name, `writeTypeCore` is confusing (what does `core` mean?), but you burned the good one just so that callers can omit a parameter.

Reply:
Hmm, this is a normal naming convention in C#--nice entrypoint vs. "the part that does the recursion" being "core". Since it's not exported anyway, I think it's reasonable to just include the param here with a doc comment saying `nil` is ok.

Or... I could use `func` local variables to keep the simple API without spilling the implementation into more methods. While I'm at it, I'll inline `writeTypeValue` because I only used it in one place, and inlining makes the recursion a little more obvious.